cli/validate_inputs.go

```diff
@@ -0,0 +1,317 @@
+			terragruntOptions.Logger.Errorf("\t- %s", varName)
+		}
+		terragruntOptions.Logger.Error("")
+	} else {
+		terragruntOptions.Logger.Info("All required inputs are passed in by terragrunt.")
+	}
+
+	// Return an error when there are misaligned inputs.
+	if len(unusedVars) > 0 || len(missingVars) > 0 {
+		return fmt.Errorf("Terragrunt configuration has misaligned inputs")
+	}
+	return nil
+}
+
+// getDefinedTerragruntInputs will return a list of names of all variables that are configured by terragrunt to be
+// passed into terraform. Terragrunt can pass in inputs from:
+// - var files defined on terraform.extra_arguments blocks.
+// - -var and -var-file args passed in on extra_arguments CLI args.
+// - env vars defined on terraform.extra_arguments blocks.
+// - env vars from the external runtime calling terragrunt.
+// - inputs blocks.
+// - automatically injected terraform vars (terraform.tfvars, terraform.tfvars.json, *.auto.tfvars, *.auto.tfvars.json)
+func getDefinedTerragruntInputs(terragruntOptions *options.TerragruntOptions, workingConfig *config.TerragruntConfig) ([]string, error) {
+	envVarTFVars := getTerraformInputNamesFromEnvVar(terragruntOptions, workingConfig)
+	inputsTFVars := getTerraformInputNamesFromConfig(workingConfig)
+	varFileTFVars, err := getTerraformInputNamesFromVarFiles(terragruntOptions, workingConfig)
+	if err != nil {
+		return nil, err
+	}
+	cliArgsTFVars, err := getTerraformInputNamesFromCLIArgs(terragruntOptions, workingConfig)
+	if err != nil {
+		return nil, err
+	}
+	autoVarFileTFVars, err := getTerraformInputNamesFromAutomaticVarFiles(terragruntOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	// Dedupe the input vars. We use a map as a set to accomplish this.
+	tmpOut := map[string]bool{}
+	for _, varName := range envVarTFVars {
+		tmpOut[varName] = true
+	}
+	for _, varName := range inputsTFVars {
+		tmpOut[varName] = true
+	}
+	for _, varName := range varFileTFVars {
+		tmpOut[varName] = true
+	}
+	for _, varName := range cliArgsTFVars {
+		tmpOut[varName] = true
+	}
+	for _, varName := range autoVarFileTFVars {
+		tmpOut[varName] = true
+	}
+
+	out := []string{}
+	for varName, _ := range tmpOut {
+		out = append(out, varName)
+	}
+	return out, nil
+}
+
+// getTerraformInputNamesFromEnvVar will check the runtime environment variables and the configured environment
+// variables from extra_arguments blocks to see if there are any TF_VAR environment variables that set terraform
+// variables. This will return the list of names of variables that are set in this way by the given terragrunt
+// configuration.
+func getTerraformInputNamesFromEnvVar(terragruntOptions *options.TerragruntOptions, terragruntConfig *config.TerragruntConfig) []string {
+	envVars := terragruntOptions.Env
+
+	// Make sure to check if there are configured env vars in the parsed terragrunt config.
+	if terragruntConfig.Terraform != nil {
+		for _, arg := range terragruntConfig.Terraform.ExtraArgs {
+			// only focus on plan args
```

Comment:
Sorry, why only plan args?

Reply:
I originally had it checking only `plan` because I thought it would simplify the code, but looking back at it now, I guess it doesn't matter to go through all the args. 90% of the time this check would be moot anyway (since most people probably configure var args using `get_terraform_commands_that_need_vars`.

Reply:
Removed check: 295ec58eae93847070e42b16aa44c6b767d47416